File: test/cd.test.js
const { describe, it, beforeEach } = require("node:test");
const assert = require("node:assert/strict");
const gitModule = require("../lib/git");
const configModule = require("../lib/config");

const WORKTREE_OUTPUT = [
  "/projects/myrepo/.bare           (bare)",
  "/projects/myrepo/main            abc1234 [main]",
  "/projects/myrepo/feature-branch  def5678 [feature-branch]",
  "/projects/myrepo/prs/123         ghi9012 [fix/login-bug]",
  "/projects/myrepo/prs/456         jkl3456 [feat/signup]",
].join("\n");

function loadCd(t) {
  t.mock.method(configModule, "loadConfig", () => ({
    root: "/projects/myrepo",
    config: {},
  }));
  t.mock.method(gitModule, "gitInBare", () => WORKTREE_OUTPUT);
  delete require.cache[require.resolve("../lib/commands/cd")];
  return require("../lib/commands/cd");
}

describe("cd", () => {
  let stdout;
  let stderr;

  beforeEach((t) => {
    stdout = "";
    stderr = "";
    t.mock.method(process.stdout, "write", (s) => {
      stdout += s;
    });
    t.mock.method(process.stderr, "write", (s) => {
      stderr += s;
    });
  });

  it("prints project root when no name given", (t) => {
    t.mock.method(configModule, "loadConfig", () => ({
      root: "/projects/myrepo",
      config: {},
    }));
    delete require.cache[require.resolve("../lib/commands/cd")];
    const cd = require("../lib/commands/cd");

    cd();

    assert.equal(stdout, "/projects/myrepo\n");
  });

  it("matches exact branch name", (t) => {
    const cd = loadCd(t);
    cd("feature-branch");
    assert.equal(stdout, "/projects/myrepo/feature-branch\n");
  });

  it("matches directory basename for PR worktrees", (t) => {
    const cd = loadCd(t);
    cd("123");
    assert.equal(stdout, "/projects/myrepo/prs/123\n");
  });

  it("matches relative path", (t) => {
    const cd = loadCd(t);
    cd("prs/123");
    assert.equal(stdout, "/projects/myrepo/prs/123\n");
  });

  it("matches substring on branch name", (t) => {
    const cd = loadCd(t);
    cd("login");
    assert.equal(stdout, "/projects/myrepo/prs/123\n");
  });

  it("exits with error when no match found", (t) => {
    const cd = loadCd(t);
    t.mock.method(process, "exit", () => {
      throw new Error("process.exit");
    });

    assert.throws(() => cd("nonexistent"), /process\.exit/);
    assert.match(stderr, /No worktree found matching "nonexistent"/);
  });

  it("exits with error when multiple matches found", (t) => {
    const cd = loadCd(t);
    t.mock.method(process, "exit", () => {
      throw new Error("process.exit");
    });

    assert.throws(() => cd("feat"), /process\.exit/);
    assert.match(stderr, /Multiple worktrees match "feat"/);
  });

  it("excludes .bare entry", (t) => {
    const cd = loadCd(t);
    t.mock.method(process, "exit", () => {
      throw new Error("process.exit");
    });

    assert.throws(() => cd(".bare"), /process\.exit/);
    assert.match(stderr, /No worktree found/);
  });
});

File: internal/cmd/cd_test.go
package cmd

import (
	"strings"
	"testing"

	"github.com/gradyholmes/wt/internal/config"
)

const testWorktreeOutput = `/projects/myrepo/.bare           (bare)
/projects/myrepo/main            abc1234 [main]
/projects/myrepo/feature-branch  def5678 [feature-branch]
/projects/myrepo/prs/123         ghi9012 [fix/login-bug]
/projects/myrepo/prs/456         jkl3456 [feat/signup]`

func cdDeps() (*MockGit, *MockConfig) {
	mg := &MockGit{
		GitInBareFunc: func(args, root string) (string, error) {
			return testWorktreeOutput, nil
		},
	}
	mc := &MockConfig{Root: "/projects/myrepo", Cfg: &config.Config{}}
	return mg, mc
}

func TestParseWorktreeList(t *testing.T) {
	entries := ParseWorktreeList(testWorktreeOutput)
	if len(entries) != 4 {
		t.Fatalf("expected 4 entries, got %d", len(entries))
	}
	// .bare should be excluded
	for _, e := range entries {
		if strings.Contains(e.Dir, ".bare") {
			t.Error("should exclude .bare entry")
		}
	}
}

func TestCdCmd(t *testing.T) {
	t.Run("prints project root when no name given", func(t *testing.T) {
		mg, mc := cdDeps()
		d, stdout, _ := testDeps(mg, mc, nil)

		cmd := CdCmd(d)
		cmd.SetArgs([]string{})
		if err := cmd.Execute(); err != nil {
			t.Fatal(err)
		}
		if got := strings.TrimSpace(stdout.String()); got != "/projects/myrepo" {
			t.Errorf("got %q, want /projects/myrepo", got)
		}
	})

	t.Run("matches exact branch name", func(t *testing.T) {
		mg, mc := cdDeps()
		d, stdout, _ := testDeps(mg, mc, nil)

		cmd := CdCmd(d)
		cmd.SetArgs([]string{"feature-branch"})
		if err := cmd.Execute(); err != nil {
			t.Fatal(err)
		}
		if got := strings.TrimSpace(stdout.String()); got != "/projects/myrepo/feature-branch" {
			t.Errorf("got %q", got)
		}
	})

	t.Run("matches directory basename for PR worktrees", func(t *testing.T) {
		mg, mc := cdDeps()
		d, stdout, _ := testDeps(mg, mc, nil)

		cmd := CdCmd(d)
		cmd.SetArgs([]string{"123"})
		if err := cmd.Execute(); err != nil {
			t.Fatal(err)
		}
		if got := strings.TrimSpace(stdout.String()); got != "/projects/myrepo/prs/123" {
			t.Errorf("got %q", got)
		}
	})

	t.Run("matches relative path", func(t *testing.T) {
		mg, mc := cdDeps()
		d, stdout, _ := testDeps(mg, mc, nil)

		cmd := CdCmd(d)
		cmd.SetArgs([]string{"prs/123"})
		if err := cmd.Execute(); err != nil {
			t.Fatal(err)
		}
		if got := strings.TrimSpace(stdout.String()); got != "/projects/myrepo/prs/123" {
			t.Errorf("got %q", got)
		}
	})

	t.Run("matches substring on branch name", func(t *testing.T) {
		mg, mc := cdDeps()
		d, stdout, _ := testDeps(mg, mc, nil)

		cmd := CdCmd(d)
		cmd.SetArgs([]string{"login"})
		if err := cmd.Execute(); err != nil {
			t.Fatal(err)
		}
		if got := strings.TrimSpace(stdout.String()); got != "/projects/myrepo/prs/123" {
			t.Errorf("got %q", got)
		}
	})

	t.Run("errors when no match found", func(t *testing.T) {
		mg, mc := cdDeps()
		d, _, stderr := testDeps(mg, mc, nil)

		cmd := CdCmd(d)
		cmd.SilenceErrors = true
		cmd.SilenceUsage = true
		cmd.SetArgs([]string{"nonexistent"})
		err := cmd.Execute()
		if err == nil {
			t.Fatal("expected error")
		}
		if !strings.Contains(stderr.String(), `No worktree found matching "nonexistent"`) {
			t.Errorf("stderr = %q", stderr.String())
		}
	})

	t.Run("errors when multiple matches found", func(t *testing.T) {
		mg, mc := cdDeps()
		d, _, stderr := testDeps(mg, mc, nil)

		cmd := CdCmd(d)
		cmd.SilenceErrors = true
		cmd.SilenceUsage = true
		cmd.SetArgs([]string{"feat"})
		err := cmd.Execute()
		if err == nil {
			t.Fatal("expected error")
		}
		if !strings.Contains(stderr.String(), `Multiple worktrees match "feat"`) {
			t.Errorf("stderr = %q", stderr.String())
		}
	})

	t.Run("excludes .bare entry", func(t *testing.T) {
		mg, mc := cdDeps()
		d, _, stderr := testDeps(mg, mc, nil)

		cmd := CdCmd(d)
		cmd.SilenceErrors = true
		cmd.SilenceUsage = true
		cmd.SetArgs([]string{".bare"})
		err := cmd.Execute()
		if err == nil {
			t.Fatal("expected error")
		}
		if !strings.Contains(stderr.String(), "No worktree found") {
			t.Errorf("stderr = %q", stderr.String())
		}
	})
}
